daemon/logger/loggerutils/logfile.go

```diff
@@ -95,43 +161,112 @@ func (w *LogFile) checkCapacityAndRotate() error {
 	}
 
 	if w.currentSize >= w.capacity {
-		name := w.f.Name()
+		w.rotateMu.Lock()
 		if err := w.f.Close(); err != nil {
+			w.rotateMu.Unlock()
 			return errors.Wrap(err, "error closing file")
 		}
-		if err := rotate(name, w.maxFiles); err != nil {
+		if err := rotate(w.logPath, w.maxFiles, w.compress); err != nil {
+			w.rotateMu.Unlock()
 			return err
 		}
-		file, err := os.OpenFile(name, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
+		file, err := os.OpenFile(w.logPath, os.O_WRONLY|os.O_TRUNC|os.O_CREATE, w.perms)
 		if err != nil {
+			w.rotateMu.Unlock()
 			return err
 		}
 		w.f = file
 		w.currentSize = 0
 		w.notifyRotate.Publish(struct{}{})
+
+		if w.maxFiles <= 1 || !w.compress {
+			w.rotateMu.Unlock()
+			return nil
+		}
+
+		go func() {
+			compressFile(w.logPath+".1", w.lastTimestamp)
```

Comment:
So, only time this could be a problem is if the next rotation takes place before compression is completed; not sure if that's a real issue, so just thinking out loud. I did some extreme test;

```patch
diff --git a/daemon/logger/loggerutils/logfile.go b/daemon/logger/loggerutils/logfile.go
index e0f21f46c..40bbba485 100644
--- a/daemon/logger/loggerutils/logfile.go
+++ b/daemon/logger/loggerutils/logfile.go
@@ -254,7 +254,7 @@ func compressFile(fileName string, lastTimestamp time.Time) {
                // Here log the error only and don't return since this is just an optimization.
                logrus.Warningf("Failed to marshal JSON: %v", err)
        }
-
+       time.Sleep(50 * time.Second)
        _, err = pools.Copy(compressWriter, file)
        if err != nil {
                logrus.WithError(err).WithField("module", "container.logs").WithField("file", fileName).Error("Error compressing log file")
```

Running with that, rotation is waiting for compression to complete, and killing the container became problematic, which is likely expected, just thinking if there would be real-world situations where that would happen.

Reply:
Yeah, this should be expected. There's possibly more we can do to prevent blocking on killing the container, but there's already a lot going on with locking for this case.

Reply:
Yup, it was probably a bit extreme, just had my "QA hat" on, and seeing what possible things could be problematic. Don't think it should be a blocker for this feature.